### PR TITLE
Refactored set access control and added builders for different types

### PR DESCRIPTION
**Pull request by gapra-msft**
Resolves https://github.com/Azure/azure-sdk-for-java/issues/6100 to create builders for the permissions/accessControlList
Resolves https://github.com/Azure/azure-sdk-for-java/issues/6113 to split setAccessControl into setAccessControlList and setPermissions 
Updates README for datalake

### CODE REVIEW

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy [hunk 1]
```diff
@@ -14,10 +14,7 @@ import com.azure.core.util.Configuration
 import com.azure.core.util.logging.ClientLogger
 import com.azure.identity.EnvironmentCredentialBuilder
 import com.azure.storage.common.StorageSharedKeyCredential
-import com.azure.storage.file.datalake.models.FileSystemItem
-import com.azure.storage.file.datalake.models.LeaseStateType
-import com.azure.storage.file.datalake.models.ListFileSystemsOptions
-import com.azure.storage.file.datalake.models.PathProperties
+import com.azure.storage.file.datalake.models.*
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: Checkstyle should be mad at this.

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    def "RolePermissions parse Symbolic"() {
+        expect:
+        RolePermissions.parseSymbolic(symbol, false).toOctal() == permission
+
+        where:
+        symbol | permission
+        "--x"  | "1"
+        "-w-"  | "2"
+        "r--"  | "4"
+    }
```
Comment (jaschrep-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy]: These tests don't seem to mirror each other in types. Why does `toOctal` produce a string and not a numeric value?

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java [hunk 3]
```diff
@@ -371,54 +374,108 @@ public DataLakeServiceVersion getServiceVersion() {
...
+     * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update">Azure Docs</a></p>
+     *
+     * @param accessControlList The access control list of the resource.
+     * @param group The group of the resource.
+     * @param owner The owner of the resource.
+     * @param accessConditions {@link DataLakeRequestConditions}
+     * @return A reactive response containing the resource info.
+     */
+    public Mono<Response<PathInfo>> setAccessControlListWithResponse(List<PathAccessControlEntry> accessControlList,
+        String group, String owner, DataLakeRequestConditions accessConditions) {
```
[the remark this answers is not in the document]
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java]: https://github.com/Azure/azure-sdk-for-java/issues/6101 This is coming soon

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy (hunk 1 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: I think checkstyle doesnt run on test files

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy (hunk 2 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy]: toOctal means convert to a string with octal representation, suggestions on a better name are welcome

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControl.java [hunk 4]
```diff
@@ -95,7 +97,7 @@ public String getPermissions() {
      * @param permissions the permissions value to set.
      * @return the PathAccessControl object itself.
      */
-    public PathAccessControl setPermissions(String permissions) {
+    public PathAccessControl setPermissions(PathPermissions permissions) {
```
[the remark this answers is not in the document]
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControl.java]: I should probably remove all setters now that users will not use this to set the pathaccesscontrol

[the remark this answers is not in the document]
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControl.java]: Removed the setters - renamed all getters as get___

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.storage.common.implementation.StorageImplUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * POSIX access control rights on files and directories. The value is a comma-separated list of access control entries.
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java]: Could this be broken into a couple paragraphs and if needed have lists to show examples? For example

```
PathAccessControlEntry represents POSIX access control rights on files and directories.

<p>
The value is a comma-separated list of access control entries, each access control entry (ACE) consists of four elements:
<ul>
  <li>Scope</li>
  <li>Type</li>
  <li>User or Group Identifier</li>
  <li>Permissions</li>
</ul>
```

Right now this is a wall of text which has key information that could be lost in its density.

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy (hunk 1 above)
Comment (alzimmermsft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/APISpec.groovy]: I believe Checkstyle is suppressed on `.groovy` files

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+ * all users not found in any of the other entries. The user or group identifier is omitted for entries of type "mask"
+ * and "other". The user or group identifier is also omitted for the owner and owning group. The permission field is a
+ * 3-character sequence where the first character is 'r' to grant read access, the second character is 'w' to grant
+ * write access, and the third character is 'x' to grant execute permission. If access is not granted, the '-' character
+ * is used to denote that the permission is denied. For example, the following ACL grants read, write, and execute
+ * rights to the file owner and john.doe@contoso, the read right to the owning group, and nothing to everyone else:
+ * "user::rwx,user:john.doe@contoso:rwx,group::r--,other::---,mask::rwx".
+ */
+public class PathAccessControlEntry {
+    public static final String SUPERUSER = "$SUPERUSER";
```
Comment (gapra-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java]: make lowercase

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/AccessControlType.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.file.datalake.models;
+
+/**
+ * Specifies the type of the {@link PathAccessControlEntry}
+ */
+public enum AccessControlType {
```
[the remark this answers is not in the document]
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/AccessControlType.java]: Changed to expandable string enum

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy (hunk 2 above)
Comment (rickle-msft) [reply sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy]: toOctalString could clarify

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java [hunk 8]
```diff
@@ -371,54 +374,108 @@ public DataLakeServiceVersion getServiceVersion() {
...
      * <p><strong>Code Samples</strong></p>
      *
-     * {@codesnippet com.azure.storage.file.datalake.DataLakePathAsyncClient.setAccessControlWithResponse#PathAccessControl-DataLakeRequestConditions}
+     * {@codesnippet com.azure.storage.file.datalake.DataLakePathAsyncClient.setPermissions#PathPermissions-String-String}
      *
      * <p>For more information, see the
      * <a href="https://docs.microsoft.com/en-us/rest/api/storageservices/datalakestoragegen2/path/update">Azure Docs</a></p>
      *
-     * @param accessControl {@link PathAccessControl}
+     * @param permissions The permissions of the resource.
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java]: I tend to try to just have the docstrings for parameters like this be just {@link PathPermissions} and let that redirect to the docs on the type. Then we don't have to worry about keeping the docs consistent everywhere the type is used. Not sure how others feel about it, though.

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java [hunk 9]
```diff
@@ -371,54 +374,108 @@ public DataLakeServiceVersion getServiceVersion() {
...
+    public Mono<PathInfo> setPermissions(PathPermissions permissions, String group, String owner) {
+        try {
+            return setPermissionsWithResponse(permissions, group, owner, null).flatMap(FluxUtil::toMono);
+        } catch (RuntimeException ex) {
+            return monoError(logger, ex);
+        }
+    }
+
+    /**
+     * Changes the permissions for a resource.
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java]: For this and setAccessControlList, can you also mention that the api can be used to update the owner or owning group.

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+ * <p>
+ * There are four ACE types:
+ * <ul>
+ *   <li>"user": grants rights to the owner or a named user</li>
+ *   <li>"group" grants rights to the owning group or a named group</li>
+ *   <li>"mask" restricts rights granted to named users and the members of groups</li>
+ *   <li>"other" grants rights to all users not found in any of the other entries</li>
+ * </ul>
+ * <p>
+ * The user or group identifier is omitted for entries of type "mask" and "other".
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java]: Can we mention here (and on constructors/setters) that the id is a AAD ObjectId?

File: sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        then:
+        permissions.other().execute() == execute
+        permissions.other().read() == read
+        permissions.other().write() == write
+        permissions.stickyBit() == stickyBit
+        permissions.extendedInfoInAcl() == extendedInfoInAcl
+
+        where:
+        // These test the value of other
+        symbol      | execute | read | write | stickyBit | extendedInfoInAcl
```
Comment (rickle-msft) [inline sdk/storage/azure-storage-file-datalake/src/test/java/com/azure/storage/file/datalake/ModelTests.groovy]: nit: spock style suggests using a || to separate inputs and expected outputs, so the column after symbol should be a column of ||

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java (hunk 8 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java]: good point. done

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java (hunk 9 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/DataLakePathAsyncClient.java]: done

File: sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java (hunk 10 above)
Comment (gapra-msft) [reply sdk/storage/azure-storage-file-datalake/src/main/java/com/azure/storage/file/datalake/models/PathAccessControlEntry.java]: Added it here. Looks like it is already specified in the getter/setter

**Pull request merged**